app/scripts/lib/crypto/recovery-keys.js

```diff
@@ -0,0 +1,175 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define(function (require, exports, module) {
+  const sjcl = require('sjcl');
+
+  function importFxaCryptoDeriver() {
+    return import(/* webpackChunkName: "fxaCryptoDeriver" */ 'fxaCryptoDeriver');
+  }
+
+  const RECOVERY_KEY_VERSION = '\x01';
+  const ALPHABET_BASE32 = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';
+
+  return {
+    /**
+     * Generate a random base32 string of length and
+     * prepend encryption version number.
+     *
+     * @param {Integer} length Length of string to generate
+     * @returns {string}
+     */
+    generateRecoveryKey: function (length) {
+      return importFxaCryptoDeriver().then((fxaCryptoDeriver) => {
+        const jose = fxaCryptoDeriver.jose;
+        const bytes = jose.util.randomBytes(length - 1);
+        const code = [];
+        for (let i = 0; i < length; i++) {
+          code.push(ALPHABET_BASE32[bytes[i] % 32]);
+        }
+
+        // Prepend version and return code
+        return RECOVERY_KEY_VERSION + code.join('');
+      });
+    },
+
+    /**
+     * Creates the encryption keys from the uid and recovery key. The object contains
+     * `recoveryKeyId` which can be used to retrieve the recovery data and
+     * `recoveryEnc` which can be used to decrypt the data.
+     *
+     * @param {String} uid Uid of user
+     * @param {String} recoveryKey Base32 encoded string
+     * @returns {Promise} A promise that will be fulfilled with the `recoveryKeyId` and `recoveryEnc`
+     */
+    getRecoveryEncryption: function (uid, recoveryKey) {
+      return Promise.resolve()
+        .then(() => {
+          if (! uid) {
+            throw new Error('Missing uid');
+          }
+
+          if (! recoveryKey) {
+            throw new Error('Missing recoveryKey');
+          }
+
+          const km = sjcl.codec.base32.toBits(recoveryKey);
+          const salt = sjcl.codec.hex.toBits(uid);
+
+          const info1 = sjcl.codec.utf8String.toBits('fxa recovery fingerprint');
+          const info2 = sjcl.codec.utf8String.toBits('fxa recovery encrypt key');
+
+          const recoveryKeyIdRaw = sjcl.misc.hkdf(km, 16 * 8, salt, info1);
+          const recoveryKeyEncRaw = sjcl.misc.hkdf(km, 32 * 8, salt, info2);
+
+          // Convert to hex strings for readability
+          const recoveryKeyId = sjcl.codec.hex.fromBits(recoveryKeyIdRaw);
+          const recoveryEnc = sjcl.codec.hex.fromBits(recoveryKeyEncRaw);
+
+          return {
```

Comment:
If turning recoveryEnc+recoveryKeyId into a JWK here, I would suggest setting the resulting JWK's `kid` to be (the hex of) recoveryKeyId:

```
return jose.JWK.asKey({
  kty: 'oct',
  k: recoveryEnc,
  alg: 'A256GCM',
  kid: recoveryKeyId
});
```